### DISCUSSION TITLE
Generation of taxonomy validation objects for CRDs

### DISCUSSION THREAD

**Discussion by roee88** [fybrik/fybrik#848] [Ideas]
Looking at the flow for taxonomy validation of CRDs (e.g., FybrikApplication and [FybrikModule](https://github.com/fybrik/fybrik/pull/804)) I'm a bit worried about keeping consistency between the structure in the CRD and the structure in the validation object. The latter must be a subset of the former with `additionalProperties: true` set in the right places to allow the additional fields that don't require validation. Whenever the structure of the CRD would change we will need to be careful about the structure of the validation object. 

I wrote a small command to generate the validation object from the CRDs based on the description of each field. If the description includes `taxonomy.json#/definitions/<some name>` then the field will be included in the validation object accordingly. The code is available in https://github.com/fybrik/fybrik/compare/master...roee88:taxonomy-crd-generator. 

I would like to get feedback on the approach. If adopted we will need to add the taxonomy reference in a field's description for each field that requires taxonomy validation and also invoke the command as part of manifests generation.

**Comment by simanadler** [comment]
@roee88 The general approach makes sense to me.  Is there a reason though to limit the generation from CRDs?  What if a connector has a structure that needs to be validated?

**Reply by roee88** [reply]
Eventually we can do that too, it's a little bit harder to implement but definitely possible. However, it's not as important right now because while CRDs are defined in the code (so we get a code->schema sync issue), connector related structures are defined just in schema (the code is generated from it so no sync issue). But eventually yes it could be nice.
